### PR TITLE
perf(test_utils): replace sleep with async queue join when stopping test stream

### PR DESCRIPTION
**Pull request by JeroennC**
TestStreamClient uses async queues in the background, which can be awaited for all processing to be done. This speeds up tests starting and stopping the TestStreamClient by removing the 1s sleep.

An assumption is made to call `task_done` in the TestConsumer.getone, as there is no callback towards the consumer after processing a record. I believe the existing test cases should still be sufficient, please let me know if you see some edge cases where this would not work or needs extra testing.

### CODE REVIEW

File: kstreams/test_utils/test_clients.py [hunk 1]
```diff
@@ -17,6 +17,8 @@ async def start(self):
 
 
 class TestProducer(Base, Producer):
+    __test__ = False
```
Comment (JeroennC) [inline kstreams/test_utils/test_clients.py:20]: Added these small changes to prevent pytest from trying to collect them as tests

File: tests/test_stream_engine.py [hunk 2]
```diff
@@ -95,6 +96,8 @@ async def stream(_):
             Consumer.start.assert_awaited()
             stream_engine._producer.start.assert_awaited()
 
+            await asyncio.sleep(0)  # Allow stream coroutine to run once
```
Comment (JeroennC) [inline tests/test_stream_engine.py:99]: Got warnings about the stream coroutine never being awaited, this was because the task would be cancelled before it got the time to run on the event loop.
This sleep passes context to the event loop to let the task with the stream coroutine in it run

File: kstreams/test_utils/test_clients.py [hunk 3]
```diff
@@ -145,6 +150,11 @@ def partitions_for_topic(self, topic: str) -> Set:
     async def getone(
         self,
     ) -> Optional[ConsumerRecord]:  # The return type must be fixed later on
+        if self._previous_topic:
+            # Assumes previous record retrieved through getone was completed
+            self._previous_topic.task_done()
```
Comment (marcosschroh) [inline kstreams/test_utils/test_clients.py:155]: I am wondering whether we can do `topic.task_done()` before `return consumer_record` instead, then we won't need `_previous_topic`. It can be argue that `task_done()` should be called only if the event was properly processed (like a commit) but I think during testing it won't make too much difference.

File: kstreams/test_utils/topics.py [hunk 4]
```diff
@@ -106,6 +112,13 @@ def all_messages_consumed(cls) -> bool:
                 return False
         return True
 
+    @classmethod
+    async def join(cls) -> None:
+        """
+        Wait for all topic messages to be processed
+        """
+        await asyncio.gather(*[topic.join() for topic in cls.topics.values()])
```
Comment (marcosschroh) [inline kstreams/test_utils/topics.py:120]: Cool!

File: kstreams/test_utils/test_clients.py (hunk 3 above)
Comment (JeroennC) [reply kstreams/test_utils/test_clients.py:155]: We could do it, but it leaves us with risks during testing that are also there currently, namely if the consumption of an event takes a long time (asynchronously), the stream will be stopped before the event is able to be processed fully.

For example in the test `test_streams_consume_events` if we add an `await asyncio.sleep(1)` before the `save_to_db(cr.value)` call, doing `task_done()` before `return consumer_record` will then get the queue.join to return instantly, stopping the server and exiting the test before the `save_to_db(cr.value)` is made.
